_posts/2020-08-26-producer-consumer-pipelines-with-system-threading-channels.markdown

```diff
@@ -0,0 +1,328 @@
+foreach (var postPath in postPaths)
+{
+    await postPathsChannel.Writer.WriteAsync(postPath);
+}
+
+postPathsChannel.Writer.TryComplete();
+```
+
+A couple of things to unwrap...
+
+* We are creating an unbounded channel of `string` values, because we expect the file paths to not exhaust memory. It's just some strings, right?
+* The unbounded channel is created for multiple readers, and a single writer. It's not required to specify these options, but it helps optimize the data access structure for scenarios with multiple readers/writers.
+* We're posting messages to the channel using `WriteAsync()`, and once all messages are posted, we complete the channel (`TryComplete()`), to let consumers know that no new data is going to come.
+
+On the consumer side, we can process this channel:
+
+```csharp
+var frontMatterChannel = Channel.CreateBounded<(string?, FrontMatter?)>(
+    new BoundedChannelOptions(10) { SingleReader = false, SingleWriter = false });
+
+while (await postPathsChannel.Reader.WaitToReadAsync())
+{
+    while (postPathsChannel.Reader.TryRead(out var postPath))
+    {
+        var frontMatter = await generator.ReadFrontMatterAsync(postPath);
+        await frontMatterChannel.Writer.WriteAsync((postPath, frontMatter));
+    }
+}
+
+frontMatterChannel.Writer.TryComplete();
+```
+
+A couple of things to unwrap...
+
+* While `postPathsChannel` has not been completed, we `TryRead()` messages.
+* If a message is present, we process the front matter. The processed front matter is then posted to the next channel, `frontMatterChannel`.
+* The `frontMatterChannel` channel is a bounded one. If there are 10 messages in there, the `Task` returned by `WriteAsync()`, by default, will not complete until capacity is available. Since we are using `await`, this will essentially "wait" for capacity, while other tasks in our program can run.
+* Similar to our previous producer, we `TryComplete()` the `frontMatterChannel` to let the next consumer know there will not be new data coming.
+
+For every step, we'll do something like the above. There are some good [examples of producer/consumer patterns](https://github.com/dotnet/corefxlab/blob/31d98a89d2e38f786303bf1e9f8ba4cf5b203b0f/src/System.Threading.Tasks.Channels/README.md#example-producerconsumer-patterns) in the .NET Core GitHub repo.
+
+If you want to have a look at the full implementation, [it's available on GitHub](https://github.com/maartenba/ChannelsExample).
+
+We're done, right? Well, yes and no...
+
+## Hello, `Open.ChannelExtensions`!
+
+After writing a few pipelines of producers/consumers, the channel setup, writing, completion, ... becomes cumbersome rather quickly.
+
+This is where [`Open.ChannelExtensions`](https://github.com/Open-NET-Libraries/Open.ChannelExtensions) comes in. It helps simplify
+working with `System.Threading.Channels`. It can read from and write to channels, pipe channels nto other channels, filter, batch, join, and more.
```

Comment:
```suggestion
working with `System.Threading.Channels`. It can read from and write to channels, pipe channels into other channels, filter, batch, join, and more.
```